diff --git a/arch/arm/dts/am335x-boneblack.dts b/arch/arm/dts/am335x-boneblack.dts
index f79a6bc8a..9a546eb37 100644
--- a/arch/arm/dts/am335x-boneblack.dts
+++ b/arch/arm/dts/am335x-boneblack.dts
@@ -18,6 +18,60 @@
 		device_type = "memory";
 		reg = <0x80000000 0x20000000>; /* 512 MB */
 	};
+
+	aliases {
+		bootstate = &bootstate;
+	};
+
+	bootstate: bootstate {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		magic = <0x41533137>;
+		compatible = "barebox,state";
+		backend-type = "raw";
+		backend = <&backend_state_mmc2>;
+		backend-stridesize = <0x40>;
+		backend-storage-type = "direct";
+
+		last_chosen@0 {
+			reg = <0x0 0x4>;
+			type = "uint32";
+		};
+
+		system0 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			remaining_attempts@4 {
+				reg = <0x4 0x4>;
+				type = "uint32";
+				default = <3>;
+			};
+
+			priority@8 {
+				reg = <0x8 0x4>;
+				type = "uint32";
+				default = <20>;
+			};
+		};
+
+		system1 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			remaining_attempts@C {
+				reg = <0xC 0x4>;
+				type = "uint32";
+				default = <3>;
+			};
+
+			priority@10 {
+				reg = <0x10 0x4>;
+				type = "uint32";
+				default = <20>;
+			};
+		};
+	};
 };
 
 &ldo3_reg {
@@ -28,6 +82,17 @@
 
 &mmc1 {
 	vmmc-supply = <&vmmcsd_fixed>;
+
+	partitions {
+		compatible = "fixed-partitions";
+		#size-cells = <1>;
+		#address-cells = <1>;
+		backend_state_mmc1: mmc1_state@0xb00000 {
+			reg = <0xb00000 0x100000>; /* 11MB offset (FAT boot partition)*/
+			label = "bootstate";
+		};
+	};
+
 };
 
 &mmc2 {
@@ -37,6 +102,16 @@
 	bus-width = <8>;
 	status = "okay";
 	ti,vcc-aux-disable-is-sleep;
+
+	partitions {
+		compatible = "fixed-partitions";
+		#size-cells = <1>;
+		#address-cells = <1>;
+		backend_state_mmc2: mmc2_state@0xb00000 {
+			reg = <0xb00000 0x100000>; /* 11MB offset (FAT boot partition)*/
+			label = "bootstate";
+		};
+	};
 };
 
 &am33xx_pinmux {
